compiler/src/dotty/tools/dotc/transform/ElimOpaque.scala

```diff
@@ -52,4 +56,18 @@ class ElimOpaque extends MiniPhase with DenotTransformer {
         ref
     }
   }
+
+  /** Resolve overloading of `==` and `!=` methods with the representation
+   *  types in order to avoid boxing. TODO: This should be in the SLS.
+   */
+  override def transformApply(tree: Apply)(using Context): Tree =
+    val sym = tree.symbol
+    if sym == defn.Any_== || sym == defn.Any_!= then
+      tree match
+        case Apply(Select(receiver, name: TermName), args) =>
+          applyOverloaded(receiver, name, args, Nil, defn.BooleanType)
+        case _ =>
+          tree
+    else
+      tree
```

Comment:
It would be nice to restrict it to opaque types. Let's see how it impacts performance.

Reply:
Yes, I agree.